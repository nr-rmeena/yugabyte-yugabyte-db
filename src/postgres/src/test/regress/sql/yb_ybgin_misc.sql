--
-- Yugabyte-owned test on ybgin index access method for things not covered
-- elsewhere.
--

-- Disable sequential scan so that index scan is always chosen.
SET enable_seqscan = off;

-- Set jsonbs to have jsonb_ops index, not jsonb_path_ops index.
DROP INDEX jsonbs_j_idx;
CREATE INDEX ON jsonbs USING ybgin (j);

--
-- DELETE/UPDATE
--

-- tsvector
INSERT INTO vectors (v) VALUES ('a b c i k'), ('abc j k');
EXPLAIN (costs off)
UPDATE vectors SET v = ts_delete(v, '{c, j}'::text[]) WHERE v @@ 'k';
UPDATE vectors SET v = ts_delete(v, '{c, j}'::text[]) WHERE v @@ 'k';
SELECT v FROM vectors WHERE v @@ 'k' ORDER BY i;
DELETE FROM vectors WHERE v @@ 'k';
-- jsonb
INSERT INTO jsonbs (j) VALUES ('[4,7,"sh",3,"du"]'), ('{"du":[7,9], "sh":13}');
EXPLAIN (costs off)
UPDATE jsonbs SET j = j - 'sh' WHERE j ? 'du';
UPDATE jsonbs SET j = j - 'sh' WHERE j ? 'du';
SELECT j FROM jsonbs WHERE j ? 'du' ORDER BY i;
DELETE FROM jsonbs WHERE j ? 'du';

--
-- Long strings.
--
DO $$
DECLARE
    c int;
    s text := '0000000000'
           || '1111111111'
           || '2222222222'
           || '3333333333'
           || '4444444444'
           || '5555555555'
           || '6666666666'
           || '7777777777'
           || '8888888888'
           || '9999999999'
           || 'aaaaaaaaaa'
           || 'bbbbbbbbbb'
           || 'cccccccccc'
           || 'dddddddddd'
           || 'eeeeeeeeee'
           || 'ffffffffff';
BEGIN
    -- tsvector_ops
    INSERT INTO vectors (v) VALUES (to_tsvector('simple', s));
    SELECT count(*) FROM vectors WHERE v @@ to_tsquery('simple', s) INTO c;
    RAISE NOTICE 'tsvector_ops count: %', c;
    DELETE FROM vectors WHERE v @@ to_tsquery('simple', s);
    -- jsonb_ops
    INSERT INTO jsonbs (j) VALUES (('"' || s || '"')::jsonb);
    SELECT count(*) FROM jsonbs WHERE j ? s INTO c;
    RAISE NOTICE 'jsonb_ops count: %', c;
    DELETE FROM jsonbs WHERE j ? s;
END $$;

--
-- Three ways to query tsvectors containing both of two specific words.
--

-- AND inside tsquery
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ 'a & bb';
SELECT * FROM vectors WHERE v @@ 'a & bb';
-- AND between '@@' ops
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ 'a' and v @@ 'bb';
SELECT * FROM vectors WHERE v @@ 'a' and v @@ 'bb';
-- INTERSECT between '@@' selects
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ 'a' INTERSECT SELECT * FROM vectors WHERE v @@ 'bb';
SELECT * FROM vectors WHERE v @@ 'a' INTERSECT SELECT * FROM vectors WHERE v @@ 'bb';

--
-- Three ways to query tsvectors containing at least one of two specific words.
--

-- OR inside tsquery
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ 'a | c';
SELECT * FROM vectors WHERE v @@ 'a | c';
-- OR between '@@' ops
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ 'a' or v @@ 'c';
SELECT * FROM vectors WHERE v @@ 'a' or v @@ 'c';
-- UNION between '@@' selects
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ 'a' UNION SELECT * FROM vectors WHERE v @@ 'c';
SELECT * FROM vectors WHERE v @@ 'a' UNION SELECT * FROM vectors WHERE v @@ 'c';

--
-- Complicated query
--

-- Top-level disjunction where the last item is simple
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ '((a & e) | (ccc & ccd)) & o';
SELECT * FROM vectors WHERE v @@ '((a & e) | (ccc & ccd)) & o';
-- Top-level disjunction where the last item is complex
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ 'o & ((a & e) | (ccc & ccd))';
SELECT * FROM vectors WHERE v @@ 'o & ((a & e) | (ccc & ccd))';
-- TODO(jason): add top-level conjunction tests

--
-- Index build without online schema change
--
CREATE INDEX NONCONCURRENTLY idx_nonconcurrent ON jsonbs USING ybgin (j jsonb_path_ops);
DROP INDEX idx_nonconcurrent;

--
-- Various array types
--
CREATE TABLE manyarrays (
    a text[],
    b int2[],
    c char[],
    d oidvector[],
    e pg_lsn[],
    f timestamptz[]);
INSERT INTO manyarrays VALUES (
    ARRAY['abc', '123'],
    ARRAY[123, -456],
    ARRAY['g', 'h', 'i'],
    ARRAY['123456 0', '4294967295', '1 2 4']::oidvector[],
    ARRAY['0/12345', 'AC/FD', '0/0']::pg_lsn[],
    ARRAY[make_timestamptz(1970, 1, 2, 3, 4, 5.6, 'UTC'),
          make_timestamptz(1971, 2, 3, 4, 5, 6.7, 'UTC')]);
CREATE INDEX NONCONCURRENTLY ON manyarrays USING ybgin (a);
CREATE INDEX NONCONCURRENTLY ON manyarrays USING ybgin (b);
CREATE INDEX NONCONCURRENTLY ON manyarrays USING ybgin (c);
CREATE INDEX NONCONCURRENTLY ON manyarrays USING ybgin (d);
CREATE INDEX NONCONCURRENTLY ON manyarrays USING ybgin (e);
CREATE INDEX NONCONCURRENTLY ON manyarrays USING ybgin (f);

--
-- Column ordering
--

-- ASC
CREATE INDEX idx_hash ON jsonbs USING ybgin (j ASC);
-- DESC
CREATE INDEX idx_hash ON jsonbs USING ybgin (j DESC);
-- HASH
CREATE INDEX idx_hash ON jsonbs USING ybgin (j HASH);
-- NULLS FIRST
CREATE INDEX idx_hash ON jsonbs USING ybgin (j NULLS FIRST);
-- NULLS LAST
CREATE INDEX idx_hash ON jsonbs USING ybgin (j NULLS LAST);

--
-- Colocation
--

-- Setup
CREATE TABLEGROUP g;
-- Colocated table and index
CREATE TABLE garrays (i serial PRIMARY KEY, a int[]) TABLEGROUP g;
INSERT INTO garrays (a) VALUES ('{11, 22}');
CREATE INDEX ON garrays USING ybgin (a) TABLEGROUP g;
INSERT INTO garrays (a) VALUES ('{22, 33}'), ('{33, 11}');
EXPLAIN
SELECT * FROM garrays WHERE a && '{11}';
SELECT * FROM garrays WHERE a && '{11}';
-- Noncolocated table and colocated index
CREATE TABLE nogarrays (i serial PRIMARY KEY, a int[]);
INSERT INTO nogarrays (a) VALUES ('{11, 22}');
CREATE INDEX ON nogarrays USING ybgin (a) TABLEGROUP g;
INSERT INTO nogarrays (a) VALUES ('{22, 33}'), ('{33, 11}');
EXPLAIN
SELECT * FROM nogarrays WHERE a && '{11}';
SELECT * FROM nogarrays WHERE a && '{11}';
-- Cleanup
DROP TABLE garrays, nogarrays;
DROP TABLEGROUP g;
