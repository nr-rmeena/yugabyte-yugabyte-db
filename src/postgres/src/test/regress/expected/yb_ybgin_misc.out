--
-- Yugabyte-owned test on ybgin index access method for things not covered
-- elsewhere.
--
-- Disable sequential scan so that index scan is always chosen.
SET enable_seqscan = off;
-- Set jsonbs to have jsonb_ops index, not jsonb_path_ops index.
DROP INDEX jsonbs_j_idx;
CREATE INDEX ON jsonbs USING ybgin (j);
--
-- DELETE/UPDATE
--
-- tsvector
INSERT INTO vectors (v) VALUES ('a b c i k'), ('abc j k');
EXPLAIN (costs off)
UPDATE vectors SET v = ts_delete(v, '{c, j}'::text[]) WHERE v @@ 'k';
                   QUERY PLAN
-------------------------------------------------
 Update on vectors
   ->  Index Scan using vectors_v_idx on vectors
         Index Cond: (v @@ '''k'''::tsquery)
(3 rows)

UPDATE vectors SET v = ts_delete(v, '{c, j}'::text[]) WHERE v @@ 'k';
SELECT v FROM vectors WHERE v @@ 'k' ORDER BY i;
        v
-----------------
 'a' 'b' 'i' 'k'
 'abc' 'k'
(2 rows)

DELETE FROM vectors WHERE v @@ 'k';
-- jsonb
INSERT INTO jsonbs (j) VALUES ('[4,7,"sh",3,"du"]'), ('{"du":[7,9], "sh":13}');
EXPLAIN (costs off)
UPDATE jsonbs SET j = j - 'sh' WHERE j ? 'du';
                  QUERY PLAN
-----------------------------------------------
 Update on jsonbs
   ->  Index Scan using jsonbs_j_idx on jsonbs
         Index Cond: (j ? 'du'::text)
(3 rows)

UPDATE jsonbs SET j = j - 'sh' WHERE j ? 'du';
SELECT j FROM jsonbs WHERE j ? 'du' ORDER BY i;
        j
-----------------
 [4, 7, 3, "du"]
 {"du": [7, 9]}
(2 rows)

DELETE FROM jsonbs WHERE j ? 'du';
--
-- Long strings.
--
DO $$
DECLARE
    c int;
    s text := '0000000000'
           || '1111111111'
           || '2222222222'
           || '3333333333'
           || '4444444444'
           || '5555555555'
           || '6666666666'
           || '7777777777'
           || '8888888888'
           || '9999999999'
           || 'aaaaaaaaaa'
           || 'bbbbbbbbbb'
           || 'cccccccccc'
           || 'dddddddddd'
           || 'eeeeeeeeee'
           || 'ffffffffff';
BEGIN
    -- tsvector_ops
    INSERT INTO vectors (v) VALUES (to_tsvector('simple', s));
    SELECT count(*) FROM vectors WHERE v @@ to_tsquery('simple', s) INTO c;
    RAISE NOTICE 'tsvector_ops count: %', c;
    DELETE FROM vectors WHERE v @@ to_tsquery('simple', s);
    -- jsonb_ops
    INSERT INTO jsonbs (j) VALUES (('"' || s || '"')::jsonb);
    SELECT count(*) FROM jsonbs WHERE j ? s INTO c;
    RAISE NOTICE 'jsonb_ops count: %', c;
    DELETE FROM jsonbs WHERE j ? s;
END $$;
NOTICE:  tsvector_ops count: 1
NOTICE:  jsonb_ops count: 1
--
-- Three ways to query tsvectors containing both of two specific words.
--
-- AND inside tsquery
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ 'a & bb';
                   QUERY PLAN
------------------------------------------------
 Index Scan using vectors_v_idx on vectors
   Index Cond: (v @@ '''a'' & ''bb'''::tsquery)
(2 rows)

SELECT * FROM vectors WHERE v @@ 'a & bb';
 i |                  v
---+--------------------------------------
 1 | 'a':1 'bb':2 'ccc':3
 2 | 'a':2 'bb':1 'e':3 'i':4 'o':5 'u':6
(2 rows)

-- AND between '@@' ops
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ 'a' and v @@ 'bb';
                              QUERY PLAN
----------------------------------------------------------------------
 Index Scan using vectors_v_idx on vectors
   Index Cond: ((v @@ '''a'''::tsquery) AND (v @@ '''bb'''::tsquery))
(2 rows)

SELECT * FROM vectors WHERE v @@ 'a' and v @@ 'bb';
ERROR:  unsupported ybgin index scan
DETAIL:  ybgin index method cannot use more than one scan key: got 2.
HINT:  Consider rewriting the query with INTERSECT and UNION.
-- INTERSECT between '@@' selects
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ 'a' INTERSECT SELECT * FROM vectors WHERE v @@ 'bb';
                                 QUERY PLAN
-----------------------------------------------------------------------------
 SetOp Intersect
   ->  Sort
         Sort Key: "*SELECT* 1".i, "*SELECT* 1".v
         ->  Append
               ->  Subquery Scan on "*SELECT* 1"
                     ->  Index Scan using vectors_v_idx on vectors
                           Index Cond: (v @@ '''a'''::tsquery)
               ->  Subquery Scan on "*SELECT* 2"
                     ->  Index Scan using vectors_v_idx on vectors vectors_1
                           Index Cond: (v @@ '''bb'''::tsquery)
(10 rows)

SELECT * FROM vectors WHERE v @@ 'a' INTERSECT SELECT * FROM vectors WHERE v @@ 'bb';
 i |                  v
---+--------------------------------------
 1 | 'a':1 'bb':2 'ccc':3
 2 | 'a':2 'bb':1 'e':3 'i':4 'o':5 'u':6
(2 rows)

--
-- Three ways to query tsvectors containing at least one of two specific words.
--
-- OR inside tsquery
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ 'a | c';
                  QUERY PLAN
-----------------------------------------------
 Index Scan using vectors_v_idx on vectors
   Index Cond: (v @@ '''a'' | ''c'''::tsquery)
(2 rows)

SELECT * FROM vectors WHERE v @@ 'a | c';
ERROR:  unsupported ybgin index scan
DETAIL:  ybgin index method cannot use more than one required scan entry: got 2.
-- OR between '@@' ops
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ 'a' or v @@ 'c';
                           QUERY PLAN
----------------------------------------------------------------
 Seq Scan on vectors
   Filter: ((v @@ '''a'''::tsquery) OR (v @@ '''c'''::tsquery))
(2 rows)

SELECT * FROM vectors WHERE v @@ 'a' or v @@ 'c';
 i |                  v
---+--------------------------------------
 1 | 'a':1 'bb':2 'ccc':3
 4 | 'a':1,2,3,4,5,6
 2 | 'a':2 'bb':1 'e':3 'i':4 'o':5 'u':6
(3 rows)

-- UNION between '@@' selects
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ 'a' UNION SELECT * FROM vectors WHERE v @@ 'c';
                              QUERY PLAN
-----------------------------------------------------------------------
 Unique
   ->  Sort
         Sort Key: vectors.i, vectors.v
         ->  Append
               ->  Index Scan using vectors_v_idx on vectors
                     Index Cond: (v @@ '''a'''::tsquery)
               ->  Index Scan using vectors_v_idx on vectors vectors_1
                     Index Cond: (v @@ '''c'''::tsquery)
(8 rows)

SELECT * FROM vectors WHERE v @@ 'a' UNION SELECT * FROM vectors WHERE v @@ 'c';
 i |                  v
---+--------------------------------------
 1 | 'a':1 'bb':2 'ccc':3
 2 | 'a':2 'bb':1 'e':3 'i':4 'o':5 'u':6
 4 | 'a':1,2,3,4,5,6
(3 rows)

--
-- Complicated query
--
-- Top-level disjunction where the last item is simple
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ '((a & e) | (ccc & ccd)) & o';
                                  QUERY PLAN
-------------------------------------------------------------------------------
 Index Scan using vectors_v_idx on vectors
   Index Cond: (v @@ '( ''a'' & ''e'' | ''ccc'' & ''ccd'' ) & ''o'''::tsquery)
(2 rows)

SELECT * FROM vectors WHERE v @@ '((a & e) | (ccc & ccd)) & o';
 i |                  v
---+--------------------------------------
 2 | 'a':2 'bb':1 'e':3 'i':4 'o':5 'u':6
(1 row)

-- Top-level disjunction where the last item is complex
EXPLAIN (costs off)
SELECT * FROM vectors WHERE v @@ 'o & ((a & e) | (ccc & ccd))';
                                  QUERY PLAN
-------------------------------------------------------------------------------
 Index Scan using vectors_v_idx on vectors
   Index Cond: (v @@ '''o'' & ( ''a'' & ''e'' | ''ccc'' & ''ccd'' )'::tsquery)
(2 rows)

SELECT * FROM vectors WHERE v @@ 'o & ((a & e) | (ccc & ccd))';
ERROR:  unsupported ybgin index scan
DETAIL:  ybgin index method cannot use more than one required scan entry: got 3.
-- TODO(jason): add top-level conjunction tests
--
-- Index build without online schema change
--
CREATE INDEX NONCONCURRENTLY idx_nonconcurrent ON jsonbs USING ybgin (j jsonb_path_ops);
DROP INDEX idx_nonconcurrent;
--
-- Various array types
--
CREATE TABLE manyarrays (
    a text[],
    b int2[],
    c char[],
    d oidvector[],
    e pg_lsn[],
    f timestamptz[]);
INSERT INTO manyarrays VALUES (
    ARRAY['abc', '123'],
    ARRAY[123, -456],
    ARRAY['g', 'h', 'i'],
    ARRAY['123456 0', '4294967295', '1 2 4']::oidvector[],
    ARRAY['0/12345', 'AC/FD', '0/0']::pg_lsn[],
    ARRAY[make_timestamptz(1970, 1, 2, 3, 4, 5.6, 'UTC'),
          make_timestamptz(1971, 2, 3, 4, 5, 6.7, 'UTC')]);
CREATE INDEX NONCONCURRENTLY ON manyarrays USING ybgin (a);
CREATE INDEX NONCONCURRENTLY ON manyarrays USING ybgin (b);
CREATE INDEX NONCONCURRENTLY ON manyarrays USING ybgin (c);
CREATE INDEX NONCONCURRENTLY ON manyarrays USING ybgin (d);
CREATE INDEX NONCONCURRENTLY ON manyarrays USING ybgin (e);
CREATE INDEX NONCONCURRENTLY ON manyarrays USING ybgin (f);
--
-- Column ordering
--
-- ASC
CREATE INDEX idx_hash ON jsonbs USING ybgin (j ASC);
ERROR:  access method "ybgin" does not support ASC/DESC/HASH options
-- DESC
CREATE INDEX idx_hash ON jsonbs USING ybgin (j DESC);
ERROR:  access method "ybgin" does not support ASC/DESC/HASH options
-- HASH
CREATE INDEX idx_hash ON jsonbs USING ybgin (j HASH);
ERROR:  access method "ybgin" does not support ASC/DESC/HASH options
-- NULLS FIRST
CREATE INDEX idx_hash ON jsonbs USING ybgin (j NULLS FIRST);
ERROR:  access method "ybgin" does not support NULLS FIRST/LAST options
-- NULLS LAST
CREATE INDEX idx_hash ON jsonbs USING ybgin (j NULLS LAST);
ERROR:  access method "ybgin" does not support NULLS FIRST/LAST options
--
-- Colocation
--
-- Setup
CREATE TABLEGROUP g;
-- Colocated table and index
CREATE TABLE garrays (i serial PRIMARY KEY, a int[]) TABLEGROUP g;
INSERT INTO garrays (a) VALUES ('{11, 22}');
CREATE INDEX ON garrays USING ybgin (a) TABLEGROUP g;
INSERT INTO garrays (a) VALUES ('{22, 33}'), ('{33, 11}');
EXPLAIN
SELECT * FROM garrays WHERE a && '{11}';
                                   QUERY PLAN
---------------------------------------------------------------------------------
 Index Scan using garrays_a_idx on garrays  (cost=0.00..4.01 rows=1000 width=36)
   Index Cond: (a && '{11}'::integer[])
(2 rows)

SELECT * FROM garrays WHERE a && '{11}';
 i |    a
---+---------
 1 | {11,22}
 3 | {33,11}
(2 rows)

-- Noncolocated table and colocated index
CREATE TABLE nogarrays (i serial PRIMARY KEY, a int[]);
INSERT INTO nogarrays (a) VALUES ('{11, 22}');
CREATE INDEX ON nogarrays USING ybgin (a) TABLEGROUP g;
INSERT INTO nogarrays (a) VALUES ('{22, 33}'), ('{33, 11}');
EXPLAIN
SELECT * FROM nogarrays WHERE a && '{11}';
                                     QUERY PLAN
-------------------------------------------------------------------------------------
 Index Scan using nogarrays_a_idx on nogarrays  (cost=0.00..4.01 rows=1000 width=36)
   Index Cond: (a && '{11}'::integer[])
(2 rows)

SELECT * FROM nogarrays WHERE a && '{11}';
 i |    a
---+---------
 1 | {11,22}
 3 | {33,11}
(2 rows)

-- Cleanup
DROP TABLE garrays, nogarrays;
DROP TABLEGROUP g;
